### PR TITLE
enhance: Add compaction task slot usage logic

### PR DESCRIPTION
**Pull request by wayblink**
#34544

### CODE REVIEW

**Comment by wayblink** [comment]
rerun ut

File: internal/datacoord/compaction.go [hunk 1]
```diff
@@ -663,17 +663,30 @@ func (c *compactionPlanHandler) checkCompaction() error {
...
+
 	for id, slots := range nodeSlots {
-		if slots > 0 && slots > maxSlots {
+		if slots >= acquireSlot && slots > maxSlots {
 			nodeID = id
 			maxSlots = slots
 		}
 	}
+	// update the input nodeSlots
+	nodeSlots[nodeID] = nodeSlots[nodeID] - acquireSlot
```
Comment (bigsheeper) [inline internal/datacoord/compaction.go:689]: we should decrease the slots after nodeID meta saved

File: internal/datacoord/compaction_task_clustering.go [hunk 2]
```diff
@@ -558,7 +566,7 @@ func (t *clusteringCompactionTask) GetLabel() string {
 }
 
 func (t *clusteringCompactionTask) NeedReAssignNodeID() bool {
-	return t.GetState() == datapb.CompactionTaskState_pipelining && t.GetNodeID() == 0
+	return t.GetState() == datapb.CompactionTaskState_pipelining && (t.GetNodeID() == 0 || t.GetNodeID() == NullNodeID)
```
Comment (bigsheeper) [inline internal/datacoord/compaction_task_clustering.go:569]: dup with pr: https://github.com/milvus-io/milvus/pull/34537

File: internal/datanode/compaction/executor.go [hunk 3]
```diff
@@ -50,8 +52,10 @@ type executor struct {
 	completedCompactor *typeutil.ConcurrentMap[int64, Compactor]                    // planID to compactor
 	completed          *typeutil.ConcurrentMap[int64, *datapb.CompactionPlanResult] // planID to CompactionPlanResult
 	taskCh             chan Compactor
-	taskSem            *semaphore.Weighted
+	taskSem            *semaphore.Weighted             // todo remove this, unify with slot logic
 	dropped            *typeutil.ConcurrentSet[string] // vchannel dropped
+	usingSlots         *atomic.Int64
```
Comment (bigsheeper) [inline internal/datanode/compaction/executor.go:57]: Do we need this?

File: internal/datanode/compaction/executor.go [hunk 4]
```diff
@@ -37,7 +39,7 @@ const (
 
 type Executor interface {
 	Start(ctx context.Context)
-	Execute(task Compactor)
+	Execute(task Compactor) (bool, error)
```
Comment (bigsheeper) [inline internal/datanode/compaction/executor.go:42]: just error?

File: internal/datacoord/compaction_task_clustering.go (hunk 2 above)
Comment (wayblink) [reply internal/datacoord/compaction_task_clustering.go:569]: yes, I learn this from your pr

File: internal/datanode/compaction/executor.go (hunk 3 above)
Comment (wayblink) [reply internal/datanode/compaction/executor.go:57]: Actually it is a double check. Executor can refuse new task even datacoord works unexpectedly.

File: internal/datanode/compaction/executor.go (hunk 4 above)
Comment (wayblink) [reply internal/datanode/compaction/executor.go:42]: yes, if slots are used up. return a DataNodeSlotExhausted error

**Pull request merged**